inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -2,33 +2,1287 @@
+            store_scalar(ptr[reg_dst_idx + offset_r * sizeof(int)], xmm_idx_r, memory::data_type::s32);
+        }
+    }
+
+    inline void swap_scalar(bool cmp_val = true) {
+        if (isa == avx512_common) {
+            if (cmp_val)
+                vcmpps(k_mask, xmm_val_l, xmm_val_r, cmp_flg);
+            else
+                vcmpps(k_mask, xmm_idx_l, xmm_idx_r, _cmp_le_os);
+
+            uni_vmovups(xmm_tmp, xmm_val_l);
+            vblendmps(xmm_val_l | k_mask, xmm_val_r, xmm_val_l);
+            vblendmps(xmm_val_r | k_mask, xmm_tmp, xmm_val_r);
+
+            uni_vmovups(xmm_tmp, xmm_idx_l);
+            vblendmps(xmm_idx_l | k_mask, xmm_idx_r, xmm_idx_l);
+            vblendmps(xmm_idx_r | k_mask, xmm_tmp, xmm_idx_r);
+        } else {
+            if (cmp_val) {
+                if (cmp_flg == _cmp_nlt_us)
+                    vcmpge_oqps(xmm_mask, xmm_val_l, xmm_val_r);
+                else
+                    vcmple_oqps(xmm_mask, xmm_val_l, xmm_val_r);
+            } else {
+                vcmple_oqps(xmm_mask, xmm_idx_l, xmm_idx_r);
+            }
+
+            uni_vmovups(xmm_tmp, xmm_val_l);
+            vblendvps(xmm_val_l, xmm_val_r, xmm_val_l, xmm_mask);
+            vblendvps(xmm_val_r, xmm_tmp, xmm_val_r, xmm_mask);
+
+            uni_vmovups(xmm_tmp, xmm_idx_l);
+            vblendvps(xmm_idx_l, xmm_idx_r, xmm_idx_l, xmm_mask);
+            vblendvps(xmm_idx_r, xmm_tmp, xmm_idx_r, xmm_mask);
+        }
+    }
+
+    inline void swap_scalar(int l, int r, bool cmp_val = true) {
+        if (isa == avx512_common) {
+            if (cmp_val)
+                vcmpps(k_mask, xmm_val(l), xmm_val(r), cmp_flg);
+            else
+                vcmpps(k_mask, xmm_idx(l), xmm_idx(r), _cmp_le_os);
+
+            uni_vmovups(xmm_tmp, xmm_val(l));
+            vblendmps(xmm_val(l) | k_mask, xmm_val(r), xmm_val(l));
+            vblendmps(xmm_val(r) | k_mask, xmm_tmp, xmm_val(r));
+
+            uni_vmovups(xmm_tmp, xmm_idx(l));
+            vblendmps(xmm_idx(l) | k_mask, xmm_idx(r), xmm_idx(l));
+            vblendmps(xmm_idx(r) | k_mask, xmm_tmp, xmm_idx(r));
+        } else {
+            if (cmp_val) {
+                if (cmp_flg == _cmp_nlt_us)
+                    vcmpge_oqps(xmm_mask, xmm_val(l), xmm_val(r));
+                else
+                    vcmple_oqps(xmm_mask, xmm_val(l), xmm_val(r));
+            } else {
+                vcmple_oqps(xmm_mask, xmm_idx(l), xmm_idx(r));
+            }
+
+            uni_vmovups(xmm_tmp, xmm_val(l));
+            vblendvps(xmm_val(l), xmm_val(r), xmm_val(l), xmm_mask);
+            vblendvps(xmm_val(r), xmm_tmp, xmm_val(r), xmm_mask);
+
+            uni_vmovups(xmm_tmp, xmm_idx(l));
+            vblendvps(xmm_idx(l), xmm_idx(r), xmm_idx(l), xmm_mask);
+            vblendvps(xmm_idx(r), xmm_tmp, xmm_idx(r), xmm_mask);
+        }
+    }
+
+    inline void load_scalar(Xmm xmm_src, const Xbyak::Address &op, memory::data_type src_dt, bool cvt_dt = true) {
```

Comment:
Technically Yes. And I had replaced all the load/store routines, yet rolled-back some of them because of performance issue. There are three kinds of load/store in topk node. (1) vector load/store; (2) tail(oftern more than 1 element) load/sote; (3) single element load/sote. I rolled-back the 3rd kind of load/store, which mainly exists in heap sort and other innermost dimension sort, because load/store_scalar has better performance for these cases.

Some related comments from Egor and Chenhu are attached:
https://github.com/openvinotoolkit/openvino/pull/6272#discussion_r661745538
https://github.com/openvinotoolkit/openvino/pull/6272#discussion_r665906625

Further discussion is always welcomed. Thanks Dmitry!

Reply:
@xuchen-intel , could you please have a try to apply changes in [PR](https://github.com/openvinotoolkit/openvino/pull/6771) to check scalar perf. load_scalar and store_scalar should have the same perf as load/store emitters of tail being 1, because after compilation, the same instruction is executed in this case and regs pools is provided.

Reply:
@chenhu-wang Hi chenhu. I've apply the changes in https://github.com/openvinotoolkit/openvino/pull/6771 to my previous revision that with all applied load/store emitters. Still the same issue. Here is a possible reason that comes to my mind. It might be that load/store emitters are not suit for "for" loops in JIT kernel. Since these "for" loops  generate plenty of repeated instructions, yet load/store emiters have much more instructions than load/store_scalar in the JIT compilation stage, which account for plenty of time cost before first execution. But it might also be a bad idea to replace all "for" loops in JIT kernel with cmp-jmp loops for two reasons. (1) It will quite damage readability, especially to replace nested "for" loops in sorting. (2) In cmp-jmp, there is no direct "int" variable (eg. offset_byte_ needed in store_emitter_context), some extra calculation is always needed. Thus I think maybe load/store_scalar should be reserved for single element load/store. Further discussion is always welcomed. Thanks Chenhu!

Reply:
@dmitry-gorokhov As Iong as I can remember, I had replaced all the load/store_scalar routines with emitters, then roll-back some of them, because applying these load/store_emitters for scalar sorting in "for" loops introduces  performance degradation. I was thinking the reason is unlooping the "for" loops in the JIT compile consumes more time(load/store_emitters comsumes much more time than load/store_scalar routines in the jit code generating stage). But after some more discussion recently, I doubt that. Because such time consuming behavior seems will only affect first inference latency, should not affect model perormance. Anyway, as a separate task, I'll try to reproduce such perf degradation on the latest code, then fully apply load/store_emitters. Thanks Dmitry!